agent/main.go

```diff
@@ -218,14 +221,19 @@ func mainImpl() {
 	var g run.Group
 
 	// Poll for new manifests every wcPollInterval.
-	if *featureInstall {
+	if *featureUpdateSelf || *featureInstallWC {
 		cancel := make(chan interface{})
 
 		g.Add(
 			func() error {
 				for {
 
-					updateAgents(cfg, cancel)
+					if *featureUpdateSelf {
+						maybeUpdateSelf(cfg, cancel)
```

Comment:
if the self update fails, we don't want to proceed to update WC agents because WC agents may depend on a new launcher agent.

Could add a return to `maybeUpdateSelf` and if it fails to update, we skip the WC update in that cycle?

Reply:
Ah, I see.

I couldn't stop myself from wondering how to improve a couple of thing here, and one things that I cannot stop thinking about is that we could store all of the flag as a configmap, and that could make some of the aspect of configuration more convenient.
For instance, if I want to add a flag here that I'd like to disable by default, it would get undone by self-update right away. If flags were in a configmap that gets create by the bootstrap program, but only if it doesn't exist already, we have much more flexibility. This would also make a good use-case for configmaps that we can blog about.

If we need to have a way of updating the configuration, we could use to separate configmaps and we could also do the updating of it in separate (simpler) process, which I seems a little more reliable to me, as right now we have a complex process that has many tasks including one that takes care of its own configuration and state.

Besides that, your comment made me think of another option, that I think is worse considering. What if we had a simple core process that executes a script (maybe Python, or another Go program, or even [Brigade](https://github.com/Azure/brigade)), which gets downloaded periodically and lives in a ConfigMap or something like that... My point is that clearly there is a part that would evolve more dynamically then the rest, and updating that should require restarting of the controlling pod and doing the whole complicated dance associated with that, risking lose of control and any error reporting.
Having suggested that, I just realised that we could also try using a CronJob that has `imagePullPolicy: Always`, and a simpler program that reports the status of that CronJob.

Just a few ideas I got while looking at this code and thinking how we could make it a bit nicer (more reliable and easy to reason about), and perhaps we could use some of the Kubernetes primitives for that.